kayenta/kayenta-integration-tests/src/test/java/com/netflix/kayenta/configuration/EmbeddedPrometheusBootstrapConfiguration.java

```diff
@@ -13,66 +13,140 @@
-            .waitingFor(prometheusWaitStrategy)
-            .withStartupTimeout(Duration.ofSeconds(30));
-    container.start();
-    Map<String, Object> env = registerEnvironment(environment, container.getMappedPort(PORT));
-    log.info("Started Prometheus server. Connection details: {}", env);
-    return container;
+  public void startPrometheusServer(ConfigurableEnvironment env) {
+    if (prometheusContainer != null && !prometheusContainer.isRunning()) {
+      prometheusContainer.start();
+
+      Map<String, Object> prometheusEnv =
+          registerEnvironment(env, prometheusContainer.getMappedPort(PROMETHEUS_INTERNAL_PORT));
+      log.info("Started Prometheus server. Connection details: {}", prometheusEnv);
+    }
   }
 
-  static Map<String, Object> registerEnvironment(ConfigurableEnvironment environment, int port) {
+  private int waitForManagementPort() {
+    int retries = 30; // wait up to 30 seconds
+    while (retries-- > 0) {
+      String managementPortStr = environment.getProperty("local.management.port");
+      if (managementPortStr != null) {
+        return Integer.parseInt(managementPortStr);
+      }
+      try {
+        Thread.sleep(1000); // wait 1 second
+      } catch (InterruptedException ignored) {
+      }
+    }
+    throw new IllegalStateException(
+        "Property 'local.management.port' not available after waiting!");
+  }
+
+  private void exposeManagementPort(int managementPort) {
+    log.info("Exposing management port {} to Testcontainers", managementPort);
+    Testcontainers.exposeHostPorts(managementPort);
+  }
+
+  private File createPrometheusConfigFile(int managementPort) {
+    try {
+      File tempFile = File.createTempFile("prometheus", ".yml");
+      try (FileWriter writer = new FileWriter(tempFile)) {
+        writer.write(
+            String.format(
+                """
+                                global:
+                                  scrape_interval: 1s
+
+                                scrape_configs:
+                                  - job_name: kayenta-test-metrics
+                                    metrics_path: /prometheus
+                                    static_configs:
+                                      - targets: ['host.testcontainers.internal:%d']
+                                """,
+                managementPort));
+      }
+      return tempFile;
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to create prometheus.yml dynamically", e);
+    }
+  }
+
+  private Map<String, Object> registerEnvironment(ConfigurableEnvironment environment, int port) {
     Map<String, Object> map = new LinkedHashMap<>();
     map.put("embedded.prometheus.port", port);
     EnvironmentUtils.registerPropertySource("embeddedPrometheusInfo", environment, map);
     return map;
   }
+
+  public void stopPrometheusContainer() {
+    if (prometheusContainer != null && prometheusContainer.isRunning()) {
```

Comment:
Do we need to check isRunning here?  Why not call stop all the time?  It's simpler code for us, and also may protect against some race/edge conditions where some exception happens and we'd like to stop a partially started container.

Reply:
removed here https://github.com/spinnaker/spinnaker/pull/7053/commits/9ab30db4a76dec4bc1afa3eb4710cca2c740f0f3